sat_vapor_pres/sat_vapor_pres_k.F90

```diff
@@ -1148,580 +2053,1478 @@ end subroutine compute_mrs_k_0d
  integer, intent(out) :: nbad
  real    :: tmp, del
  integer :: ind, i, j, k
+ logical :: valid_types = .false.
+
+   select type (temp)
+   type is (real(kind=r4_kind))
+     select type (desat)
+     type is (real(kind=r4_kind))
+       valid_types = .true.
+     end select
+   type is (real(kind=r8_kind))
+     select type (desat)
+     type is (real(kind=r8_kind))
+       valid_types = .true.
+     end select
+   end select
+   if ( .not. valid_types ) then
+     call error_mesg ('sat_vapor_pres_k_mod::lookup_des_k_3d', 'unsupported kind', FATAL)
+   end if
 
    nbad = 0
-   do k = 1, size(temp,3)
-   do j = 1, size(temp,2)
-   do i = 1, size(temp,1)
-     tmp = temp(i,j,k)-tminl
-     ind = int(dtinvl*(tmp+tepsl))
-     if (ind < 0 .or. ind >= table_siz)  then
-       nbad = nbad+1
-     else
-       del = tmp-dtres*real(ind)
-       desat(i,j,k) = DTABLE(ind+1) + 2.*del*D2TABLE(ind+1)
-     endif
-   enddo
-   enddo
-   enddo
+
+   select type (temp)
+   type is (real(kind=r4_kind))
+     select type (desat)
+     type is (real(kind=r4_kind))
+       do k = 1, size(temp,3)
+       do j = 1, size(temp,2)
+       do i = 1, size(temp,1)
+         tmp = temp(i,j,k)-tminl
+         ind = int(dtinvl*(tmp+tepsl))
+         if (ind < 0 .or. ind >= table_siz)  then
+           nbad = nbad+1
+         else
+           del = tmp-dtres*real(ind)
+           desat(i,j,k) = DTABLE(ind+1) + 2.*del*D2TABLE(ind+1)
+         endif
+       enddo
+       enddo
+       enddo
+     end select
+   type is (real(kind=r8_kind))
+     select type (desat)
+     type is (real(kind=r8_kind))
+       do k = 1, size(temp,3)
+       do j = 1, size(temp,2)
+       do i = 1, size(temp,1)
+         tmp = temp(i,j,k)-tminl
+         ind = int(dtinvl*(tmp+tepsl))
+         if (ind < 0 .or. ind >= table_siz)  then
+           nbad = nbad+1
+         else
+           del = tmp-dtres*real(ind)
+           desat(i,j,k) = DTABLE(ind+1) + 2.*del*D2TABLE(ind+1)
+         endif
+       enddo
+       enddo
+       enddo
+     end select
+   end select
 
  end subroutine lookup_des_k_3d
 
 !#######################################################################
  subroutine lookup_des_k_2d(temp, desat, nbad)
- real, intent(in),  dimension(:,:)  :: temp
- real, intent(out), dimension(:,:)  :: desat
+ class(*), intent(in),  dimension(:,:)  :: temp
+ class(*), intent(out), dimension(:,:)  :: desat
  integer, intent(out) :: nbad
  real    :: tmp, del
  integer :: ind, i, j
+ logical :: valid_types = .false.
+
+   select type (temp)
+   type is (real(kind=r4_kind))
+     select type (desat)
+     type is (real(kind=r4_kind))
+       valid_types = .true.
+     end select
+   type is (real(kind=r8_kind))
+     select type (desat)
+     type is (real(kind=r8_kind))
+       valid_types = .true.
+     end select
+   end select
+   if ( .not. valid_types ) then
+     call error_mesg ('sat_vapor_pres_k_mod::lookup_des_k_2d', 'unsupported kind', FATAL)
```

Comment:
This error message should be more descriptive.  *"temp and desat types do not match or desat is not one of the supported types of real(kind=4) or real(kind=8)."*